File: UniversalBre/parser.cpp
#include "parser.h"
#include "op_types.h"
#include "binop_expression_node.h"

core::expression_node_ptr_s core::parser::parse()
{
    log_object_->log_debug(L"Parse program");
    // TODO: This is wrong
    parse_program();
    // TODO: This is worse than above
    return nullptr;
}

void core::parser::reset()
{
    location_ = 0;
}

core::token_type core::parser::lookahead() const
{
    return tokens_->at(location_).get_type();
}

core::token core::parser::get_token() const
{
    return tokens_->at(location_);
}

void core::parser::eat_token(
    const token_type actual,
    const token_type expected)
{
    auto actual_str_type = get_token_type_string(actual);
    auto expected_str_type = get_token_type_string(expected);

    if (actual == expected)
        log_object_->log_success(L"Matched " + actual_str_type + L" with " + expected_str_type);

    if (actual != expected)
    {
        throw exceptions::parse_failure(
            utility::wstring_to_wcstr(actual_str_type),
            utility::wstring_to_wcstr(expected_str_type));
    }
    location_++;
}

core::statement_node_vecptrptr_s core::parser::parse_program()
{
    log_object_->log_debug(L"Parse program");

    auto statements =
        utility::make_ptr_s(statement_node_ptr_vec_s());

    while (lookahead() != token_type::end_of_file)
    {
        statements->push_back(parse_possible_statement());
    }

    eat_token(lookahead(), token_type::end_of_file);
    return statements;
}

core::statement_node_ptr_s core::parser::parse_possible_statement()
{
    switch (lookahead()) {
    case token_type::identifier:
    {
        const auto result = parse_assignment_statement();
        return result;
    }
    case token_type::infinite_keyword:
    {
        return nullptr;
    }
    default:
        throw exceptions::parse_failure(L"Blah"); // TODO: FIX
    }
}

void core::parser::print_expression(expression_node_ptr_s& expression) const
{
    log_object_->log_debug(L"AST side-view");
    expression->print(0);
}

core::assignment_node_ptr_s core::parser::parse_assignment_statement()
{
    log_object_->log_debug(L"Parse function/var assignment statement");

    const auto id_name = get_token().get_lexeme();
    eat_token(lookahead(), token_type::identifier);
    eat_token(lookahead(), token_type::assignment_operator);

    switch (lookahead())
    {
    case token_type::function_keyword:
    {
        const auto function_expression = parse_function_expression();
        auto result = utility::make_ptr_s(assignment_node(id_name, function_expression));
        eat_token(lookahead(), token_type::semicolon);
        return result;
    }
    default:
    {
        auto expression = parse_expression();
        print_expression(expression);
        auto result = utility::make_ptr_s(assignment_node(id_name, expression));
        eat_token(lookahead(), token_type::semicolon);
        return result;
    }
    }
}

core::function_expression_node_ptr_s core::parser::parse_function_expression()
{
    auto result_expression = utility::make_ptr_s(function_expression_node());

    eat_token(lookahead(), token_type::function_keyword);
    if (lookahead() == token_type::left_parenthesis)
    {
        auto argument_list = parse_argument_list();
        result_expression->get_argument_list();
    }

    eat_token(lookahead(), token_type::scope_begin_operator);
    while (lookahead() != token_type::scope_end_operator)
    {
        auto assignment = parse_possible_statement();
        // TODO: Add to function expression node
    }

    eat_token(lookahead(), token_type::scope_end_operator);
    return result_expression;
}

core::argument_list_node_ptr_s core::parser::parse_argument_list()
{
    eat_token(lookahead(), token_type::left_parenthesis);
    auto argument_list = utility::make_ptr_s(argument_list_node());
    while (lookahead() == token_type::identifier)
    {
        const auto cur_token = utility::make_ptr_s(get_token());
        eat_token(lookahead(), token_type::identifier);
        argument_list->add_argument(cur_token);
        if (lookahead() != token_type::right_parenthesis)
        {
            eat_token(lookahead(), token_type::list_delimiter);
        }
    }
    eat_token(lookahead(), token_type::right_parenthesis);
    return argument_list;
}

core::expression_node_ptr_s core::parser::parse_expression()
{
    log_object_->log_debug(L"Parse expression");
    auto result = parse_boolean_or_expression();
    return result;
}

core::expression_node_ptr_s core::parser::parse_boolean_or_expression()
{
    log_object_->log_debug(L"Parse boolean or expression");
    auto left_node = parse_boolean_and_expression();
    while (lookahead() == token_type::boolean_or_operator)
    {
        auto type = op_type::op_invalid;
        switch (lookahead())
        {
        case token_type::boolean_or_operator:
            eat_token(lookahead(), token_type::boolean_or_operator);
            type = op_type::op_boolean_or;
            break;
        default: break;
        }
        const auto right_node = parse_boolean_and_expression();
        left_node = utility::make_ptr_s(
            binop_expression_node(left_node, right_node, type));
    }
    return left_node;
}

core::expression_node_ptr_s core::parser::parse_boolean_and_expression()
{
    log_object_->log_debug(L"Parse boolean and expression");
    auto left_node = parse_boolean_comparison_expression();
    while (lookahead() == token_type::boolean_and_operator)
    {
        auto type = op_type::op_invalid;
        switch (lookahead())
        {
        case token_type::boolean_and_operator:
            eat_token(lookahead(), token_type::boolean_and_operator);
            type = op_type::op_boolean_and;
            break;
        default: break;
        }
        const auto right_node = parse_boolean_comparison_expression();
        left_node = utility::make_ptr_s(
            binop_expression_node(left_node, right_node, type));
    }
    return left_node;
}

core::expression_node_ptr_s core::parser::parse_boolean_comparison_expression()
{
    log_object_->log_debug(L"Parse boolean comparison expression");
    auto left_node = parse_boolean_equality_expression();
    while (lookahead() == token_type::boolean_gt_operator
        || lookahead() == token_type::boolean_lt_operator
        || lookahead() == token_type::boolean_gte_operator
        || lookahead() == token_type::boolean_lte_operator)
    {
        auto type = op_type::op_invalid;
        switch (lookahead())
        {
        case token_type::boolean_gt_operator:
            eat_token(lookahead(), token_type::boolean_gt_operator);
            type = op_type::op_boolean_gt;
            break;
        case token_type::boolean_lt_operator:
            eat_token(lookahead(), token_type::boolean_lt_operator);
            type = op_type::op_boolean_lt;
            break;
        case token_type::boolean_lte_operator:
            eat_token(lookahead(), token_type::boolean_lte_operator);
            type = op_type::op_boolean_lte;
            break;
        case token_type::boolean_gte_operator:
            eat_token(lookahead(), token_type::boolean_gte_operator);
            type = op_type::op_boolean_gte;
            break;
        default: break;
        }
        const auto right_node = parse_boolean_equality_expression();
        left_node = utility::make_ptr_s(
            binop_expression_node(left_node, right_node, type));
    }
    return left_node;
}

core::expression_node_ptr_s core::parser::parse_boolean_equality_expression()
{
    log_object_->log_debug(L"Parse boolean equality expression");
    auto left_node = parse_concat_expression();
    while (lookahead() == token_type::boolean_eq_operator
        || lookahead() == token_type::boolean_ne_operator)
    {
        auto type = op_type::op_invalid;
        switch (lookahead())
        {
        case token_type::boolean_eq_operator:
            eat_token(lookahead(), token_type::boolean_eq_operator);
            type = op_type::op_boolean_eq;
            break;
        case token_type::boolean_ne_operator:
            eat_token(lookahead(), token_type::boolean_ne_operator);
            type = op_type::op_boolean_ne;
            break;
        default: break;
        }
        const auto right_node = parse_concat_expression();
        left_node = utility::make_ptr_s(
            binop_expression_node(left_node, right_node, type));
    }
    return left_node;
}

core::expression_node_ptr_s core::parser::parse_concat_expression()
{
    log_object_->log_debug(L"Parse concat expression");
    auto left_node = parse_math_expression();
    while (lookahead() == token_type::concat_operator)
    {
        auto type = op_type::op_invalid;
        switch (lookahead())
        {
        case token_type::concat_operator:
            eat_token(lookahead(), token_type::concat_operator);
            type = op_type::op_concat_strings;
            break;
        default: break;
        }
        const auto right_node = parse_math_expression();
        left_node = utility::make_ptr_s(
            binop_expression_node(left_node, right_node, type));
    }
    return left_node;
}

core::expression_node_ptr_s core::parser::parse_math_expression()
{
    log_object_->log_debug(L"Parse mathematical expression");
    auto left_node = parse_term();
    while (lookahead() == token_type::plus_operator
        || lookahead() == token_type::minus_operator)
    {
        auto type = op_type::op_invalid;
        switch (lookahead())
        {
        case token_type::plus_operator:
            eat_token(lookahead(), token_type::plus_operator);
            type = op_type::op_addition;
            break;
        case token_type::minus_operator:
            type = op_type::op_subtraction;
            eat_token(lookahead(), token_type::minus_operator);
            break;
        default: break;
        }
        const auto right_node = parse_term();
        left_node = utility::make_ptr_s(
            binop_expression_node(left_node, right_node, type));
    }
    return left_node;
}

core::expression_node_ptr_s core::parser::parse_term()
{
    log_object_->log_debug(L"Parse term expression");
    auto left_node = parse_factor();
    while (lookahead() == token_type::multiply_operator
        || lookahead() == token_type::divide_operator)
    {
        auto type = op_type::op_invalid;
        switch (lookahead())
        {
        case token_type::multiply_operator:
            eat_token(lookahead(), token_type::multiply_operator);
            type = op_type::op_multiplication;
            break;
        case token_type::divide_operator:
            type = op_type::op_division;
            eat_token(lookahead(), token_type::divide_operator);
            break;
        default: break;
        }
        const auto right_node = parse_factor();
        left_node = utility::make_ptr_s(
            binop_expression_node(left_node, right_node, type));
    }
    return left_node;
}

core::expression_node_ptr_s core::parser::parse_factor()
{
    log_object_->log_debug(L"Parse factor/sub-expression");
    switch (lookahead())
    {
    case token_type::left_parenthesis:
    {
        eat_token(lookahead(), token_type::left_parenthesis);
        auto single = parse_expression();
        eat_token(lookahead(), token_type::right_parenthesis);
        return single;
    }
    case token_type::integer_literal:
    {
        auto cur_lexeme = get_token().get_lexeme();
        eat_token(lookahead(), token_type::integer_literal);
        return utility::make_ptr_s(literal_expression_node(cur_lexeme));
    }
    case token_type::float_literal:
    {
        auto cur_lexeme = get_token().get_lexeme();
        eat_token(lookahead(), token_type::float_literal);
        return utility::make_ptr_s(literal_expression_node(cur_lexeme));
    }
    case token_type::identifier:
    {
        auto cur_lexeme = get_token().get_lexeme();
        eat_token(lookahead(), token_type::identifier);
        // TODO: Allow parsing of function calls
        return utility::make_ptr_s(literal_expression_node(cur_lexeme));
    }
    case token_type::string_literal:
    {
        // The grammar can allow this kind of interleaved expression
        // however, semantic analysis should provide errors on invalid operations
        // even though concatenation makes sense in this case
        auto cur_lexeme = get_token().get_lexeme();
        eat_token(lookahead(), token_type::string_literal);
        return utility::make_ptr_s(literal_expression_node(cur_lexeme));
    }
    case token_type::minus_operator:
    {
        eat_token(lookahead(), token_type::minus_operator);

        // TODO: Fix this shit, we're appending contextual information in the parser
        // as if it is scanning... dreadful.
        auto cur_lexeme = L"-" + get_token().get_lexeme();

        switch (lookahead())
        {
        case token_type::integer_literal:
            eat_token(lookahead(), token_type::integer_literal);
            return utility::make_ptr_s(literal_expression_node(cur_lexeme));
            break;
        case token_type::float_literal:
            eat_token(lookahead(), token_type::float_literal);
            return utility::make_ptr_s(literal_expression_node(cur_lexeme));
            break;
        case token_type::identifier:
            eat_token(lookahead(), token_type::identifier);
            return utility::make_ptr_s(literal_expression_node(cur_lexeme));
            break;
        default:
            auto error_str = L"Negative subexpression not allowed for " + cur_lexeme;
            throw exceptions::parse_failure(error_str.c_str());
        }
    }
    default:
        throw exceptions::parse_failure(
            L"Subexpression did not start with id, left parenthesis, identifier, or numeric constant");
    }
}

core::parser::parser(token_vecptr_s tokens, log_ptr_s log_object) :
    tokens_(std::move(tokens)),
    log_object_(std::move(log_object))
{
}

File: RustCode/src/parser.rs
use error::CompileError;
use log;
use semantic_blocks::{ArgumentBlock, DataType, ExprNode, FunctionBlock, LoopBlock, LoopType,
                      OperationType, Program, SemanticBlock, StatementBlock};
use token::{Token, TokenType};

pub struct Parser {
	location: usize,
	tokens: Vec<Token>,
}

impl Parser {
	pub fn init(tokens: Vec<Token>) -> Parser {
		Parser {
			location: 0,
			tokens: tokens,
		}
	}

	fn get_lookahead(&self) -> TokenType {
		self.tokens[self.location].get_token_type().clone()
	}

	fn eat_lookahead(&mut self, token_type: TokenType) -> Option<CompileError> {
		let lookahead = self.get_lookahead();
		self.eat_token(lookahead, token_type)
	}

	fn get_token(&self) -> Token {
		self.tokens[self.location].clone()
	}

	fn get_compile_error(&self, message: String) -> CompileError {
		let error_token = self.get_token();
		CompileError::init(0, error_token.get_line(), error_token.get_column(), message)
	}

	fn get_static_compile_error(&self, message: &'static str) -> CompileError {
		self.get_compile_error(message.to_string())
	}

	fn eat_token(&mut self, actual: TokenType, expected: TokenType) -> Option<CompileError> {
		let compile_error: CompileError;
		if actual == expected {
			let debug_message = format!("{:?} found", expected);
			log::log_debug(&debug_message);
			return None
		} else {
			let error_message = format!("Expected {:?} but got {:?}", expected, actual);
			compile_error = self.get_compile_error(error_message);
		}
		self.location += 1;
		Some(compile_error)
	}

	pub fn parse(&mut self) -> Result<Program, CompileError> {
		log::log_debug("Parse program");
		let mut semantic_blocks = Vec::<SemanticBlock>::new();
		let mut matched = true;
		while self.get_lookahead() != TokenType::EndOfFile {
			match self.get_lookahead() {
				TokenType::Identifier => {
					let assignment_statement = match self.parse_assignment_statement() {
						Ok(assignment_statement_body) => assignment_statement_body,
						Err(contents) => return Err(contents)
					};
					semantic_blocks.push(SemanticBlock::init_with_statement(assignment_statement));
				}
				TokenType::InfiniteKeyword => {
					let infinite_loop = match self.parse_infinite_loop() {
						Ok(infinite_loop_body) => infinite_loop_body,
						Err(contents) => return Err(contents)
					};
					semantic_blocks.push(SemanticBlock::init_with_loop(infinite_loop));
				}
				TokenType::FunctionKeyword => {
					let function_block = match self.parse_function_block() {
						Ok(function_block) => function_block,
						Err(contents) => return Err(contents)
					};
					semantic_blocks.push(SemanticBlock::init_with_function(function_block));
				}
				_ => {
					matched = false;
				}
			}

			if !matched {
				return Err(self.get_static_compile_error(
					"Cannot start with this statement type",
				));
			}
		}

		self.eat_lookahead(TokenType::EndOfFile);
		Ok(Program::init(semantic_blocks))
	}

	fn parse_assignment_statement(&mut self) -> Result<StatementBlock, CompileError> {
		log::log_debug("Parse assignment statement");
		let id = if self.get_lookahead() == TokenType::Identifier {
			self.get_token().get_lexeme()
		} else {
			String::new() // TODO: What?
		};
		self.eat_lookahead(TokenType::Identifier);

		let expression = match self.parse_expression() {
			Ok(expression_contents) => expression_contents,
			Err(contents) => return Err(contents)
		};
		self.eat_lookahead(TokenType::Semicolon);

		Ok(StatementBlock::init_with_assignment(id, expression))
	}

	fn parse_infinite_loop(&mut self) -> Result<LoopBlock, CompileError> {
		log::log_debug("Parse infinite loop");
		self.eat_lookahead(TokenType::Semicolon);
		Ok(LoopBlock::init(LoopType::InfiniteLoop))
	}

	fn parse_argument_list(&mut self) -> Result<Vec<ArgumentBlock>, CompileError> {
		let mut argument_list = Vec::<ArgumentBlock>::new();
		if self.get_lookahead() == TokenType::LeftParenthesis {
			self.eat_lookahead(TokenType::LeftParenthesis);
			while self.get_lookahead() == TokenType::Identifier {
				let id = self.get_token().get_lexeme();
				self.eat_lookahead(TokenType::Identifier);
				argument_list.push(ArgumentBlock::init(id));

				if self.get_lookahead() != TokenType::RightParenthesis {
					self.eat_lookahead(TokenType::ListDelimiter);
				}
			}
			self.eat_lookahead(TokenType::RightParenthesis);
		}
		Ok(argument_list)
	}

	fn parse_function_block(&mut self) -> Result<FunctionBlock, CompileError> {
		log::log_debug("Parse function");

		self.eat_lookahead(TokenType::FunctionKeyword);
		let id = if self.get_lookahead() == TokenType::Identifier {
			self.get_token().get_lexeme()
		} else {
			String::new() // TODO: Still... what?
		};

		self.eat_lookahead(TokenType::Identifier);
		let argument_list = match self.parse_argument_list() {
			Ok(argument_list_body) => argument_list_body,
			Err(contents) => return Err(contents)
		};

		let mut function_body = Vec::<SemanticBlock>::new();
		self.eat_lookahead(TokenType::ScopeBeginOperator);
		while self.get_lookahead() != TokenType::ScopeEndOperator {
			match self.parse_assignment_statement() {
				Ok(result_statement_body) => { 
					function_body.push(SemanticBlock::init_with_statement(result_statement_body)) 
				},
				Err(contents) => {
					return Err(contents)
				}
			};
		}
		self.eat_lookahead(TokenType::ScopeEndOperator);

		Ok(FunctionBlock::init(id, argument_list, function_body))
	}

	fn parse_expression(&mut self) -> Result<ExprNode, CompileError> {
		log::log_debug("Parse expression");
		self.parse_boolean_or_expression()
	}

	fn parse_boolean_or_expression(&mut self) -> Result<ExprNode, CompileError> {
		log::log_debug("Parse boolean-or subexpr");
		let mut left_node = match self.parse_boolean_and_expression() {
			Ok(left_node_unwrap) => left_node_unwrap,
			Err(contents) => return Err(contents)
		};

		while self.get_lookahead() == TokenType::BooleanOrOperator {
			match self.get_lookahead() {
				TokenType::BooleanOrOperator => {
					self.eat_lookahead(TokenType::BooleanOrOperator);
				}
				_ => break,
			}
			let right_node = match self.parse_boolean_and_expression() {
				Ok(right_node_unwrap) => right_node_unwrap,
				Err(contents) => return Err(contents)
			};
			left_node =
				ExprNode::init_as_binary(left_node, right_node, OperationType::BooleanOrOperation);
		}
		Ok(left_node)
	}

	fn parse_boolean_and_expression(&mut self) -> Result<ExprNode, CompileError> {
		log::log_debug("Parse boolean-and subexpr");
		let mut left_node = match self.parse_boolean_comparison_expression() {
			Ok(left_node_unwrap) => left_node_unwrap,
			Err(contents) => return Err(contents)
		};

		while self.get_lookahead() == TokenType::BooleanAndOperator {
			match self.get_lookahead() {
				TokenType::BooleanAndOperator => {
					self.eat_lookahead(TokenType::BooleanAndOperator);
				}
				_ => break,
			}
			let right_node = match self.parse_boolean_comparison_expression() {
				Ok(right_node_unwrap) => right_node_unwrap,
				Err(contents) => return Err(contents)
			};
			left_node =
				ExprNode::init_as_binary(left_node, right_node, OperationType::BooleanAndOperation);
		}
		Ok(left_node)
	}

	fn parse_boolean_comparison_expression(&mut self) -> Result<ExprNode, CompileError> {
		log::log_debug("Parse boolean comparison subexpr");
		let mut left_node = match self.parse_boolean_equality_expression() {
			Ok(left_node_unwrap) => left_node_unwrap,
			Err(contents) => return Err(contents)
		};

		while self.get_lookahead() == TokenType::BooleanGtOperator
			|| self.get_lookahead() == TokenType::BooleanGteOperator
			|| self.get_lookahead() == TokenType::BooleanLtOperator
			|| self.get_lookahead() == TokenType::BooleanLteOperator
		{
			let operation_type: OperationType;
			match self.get_lookahead() {
				TokenType::BooleanGtOperator => {
					self.eat_lookahead(TokenType::BooleanGtOperator);
					operation_type = OperationType::BooleanGtOperation;
				}
				TokenType::BooleanGteOperator => {
					self.eat_lookahead(TokenType::BooleanGteOperator);
					operation_type = OperationType::BooleanGteOperation;
				}
				TokenType::BooleanLtOperator => {
					self.eat_lookahead(TokenType::BooleanLtOperator);
					operation_type = OperationType::BooleanLtOperation;
				}
				TokenType::BooleanLteOperator => {
					self.eat_lookahead(TokenType::BooleanLteOperator);
					operation_type = OperationType::BooleanLteOperation;
				}
				_ => break,
			}
			let right_node = match self.parse_boolean_equality_expression() {
				Ok(right_node_unwrap) => right_node_unwrap,
				Err(contents) => return Err(contents)
			};
			left_node = ExprNode::init_as_binary(left_node, right_node, operation_type);
		}
		Ok(left_node)
	}

	fn parse_boolean_equality_expression(&mut self) -> Result<ExprNode, CompileError> {
		log::log_debug("Parse boolean equality subexpr");
		let mut left_node = match self.parse_concat_expression() {
			Ok(left_node_unwrap) => left_node_unwrap,
			Err(contents) => return Err(contents)
		};

		while self.get_lookahead() == TokenType::BooleanEqOperator
			|| self.get_lookahead() == TokenType::BooleanNeOperator
		{
			let operation_type: OperationType;
			match self.get_lookahead() {
				TokenType::BooleanEqOperator => {
					self.eat_lookahead(TokenType::BooleanEqOperator);
					operation_type = OperationType::BooleanEqOperation;
				}
				TokenType::BooleanNeOperator => {
					self.eat_lookahead(TokenType::BooleanNeOperator);
					operation_type = OperationType::BooleanNeOperation;
				}
				_ => break,
			}
			let right_node = match self.parse_concat_expression() {
				Ok(right_node_unwrap) => right_node_unwrap,
				Err(contents) => return Err(contents)
			};
			left_node = ExprNode::init_as_binary(left_node, right_node, operation_type);
		}
		Ok(left_node)
	}

	fn parse_concat_expression(&mut self) -> Result<ExprNode, CompileError> {
		log::log_debug("Parse concat subexpr");
		let mut left_node = match self.parse_mathematical_expression() {
			Ok(left_node_unwrap) => left_node_unwrap,
			Err(contents) => return Err(contents)
		};

		while self.get_lookahead() == TokenType::ConcatOperator {
			match self.get_lookahead() {
				TokenType::ConcatOperator => {
					self.eat_lookahead(TokenType::ConcatOperator);
				}
				_ => break,
			}
			let right_node = match self.parse_mathematical_expression() {
				Ok(right_node_unwrap) => right_node_unwrap,
				Err(contents) => return Err(contents)
			};

			left_node =
				ExprNode::init_as_binary(left_node, right_node, OperationType::ConcatOperation);
		}
		Ok(left_node)
	}

	fn parse_mathematical_expression(&mut self) -> Result<ExprNode, CompileError> {
		log::log_debug("Parse mathematical subexpr");

		let mut left_node = match self.parse_term() {
			Ok(left_node_unwrap) => left_node_unwrap,
			Err(contents) => return Err(contents)
		};

		while self.get_lookahead() == TokenType::PlusOperator
			|| self.get_lookahead() == TokenType::MinusOperator
		{
			let operation_type: OperationType;
			match self.get_lookahead() {
				TokenType::PlusOperator => {
					self.eat_lookahead(TokenType::PlusOperator);
					operation_type = OperationType::AdditionOperation;
				}
				TokenType::MinusOperator => {
					self.eat_lookahead(TokenType::MinusOperator);
					operation_type = OperationType::SubtractionOperation;
				}
				_ => break,
			}

			let right_node = match self.parse_term() {
				Ok(right_node_unwrap) => right_node_unwrap,
				Err(contents) => return Err(contents)
			};

			left_node = ExprNode::init_as_binary(left_node, right_node, operation_type);
		}
		Ok(left_node)
	}

	fn parse_term(&mut self) -> Result<ExprNode, CompileError> {
		log::log_debug("Parse term subexpr");
		let mut left_node = match self.parse_factor() {
			Ok(left_node_unwrap) => left_node_unwrap,
			Err(contents) => return Err(contents)
		};

		while self.get_lookahead() == TokenType::MultiplyOperator
			|| self.get_lookahead() == TokenType::DivideOperator
		{
			let operation_type: OperationType;
			match self.get_lookahead() {
				TokenType::MultiplyOperator => {
					self.eat_lookahead(TokenType::MultiplyOperator);
					operation_type = OperationType::MultiplicationOperation;
				}
				TokenType::DivideOperator => {
					self.eat_lookahead(TokenType::DivideOperator);
					operation_type = OperationType::DivisionOperation;
				}
				_ => break,
			}
			let right_node = match self.parse_factor() {
				Ok(right_node_unwrap) => right_node_unwrap,
				Err(contents) => return Err(contents)
			};
			left_node = ExprNode::init_as_binary(left_node, right_node, operation_type);
		}
		Ok(left_node)
	}

	fn parse_factor(&mut self) -> Result<ExprNode, CompileError> {
		log::log_debug("Parse factor subexpr");
		match self.get_lookahead() {
			TokenType::LeftParenthesis => {
				self.eat_lookahead(TokenType::LeftParenthesis);
				let inner_expression = self.parse_expression();
				self.eat_lookahead(TokenType::RightParenthesis);
				inner_expression
			}
			TokenType::IntegerLiteral => {
				let current_lexeme = self.get_token().get_lexeme();
				self.eat_lookahead(TokenType::IntegerLiteral);
				Ok(ExprNode::init_as_literal(
					current_lexeme,
					DataType::IntegerType,
				))
			}
			TokenType::FloatLiteral => {
				let current_lexeme = self.get_token().get_lexeme();
				self.eat_lookahead(TokenType::FloatLiteral);
				Ok(ExprNode::init_as_literal(
					current_lexeme,
					DataType::FloatType,
				))
			}
			TokenType::Identifier => {
				let current_lexeme = self.get_token().get_lexeme();
				self.eat_lookahead(TokenType::Identifier);
				Ok(ExprNode::init_as_variable(current_lexeme))
			}
			TokenType::StringLiteral => {
				let current_lexeme = self.get_token().get_lexeme();
				self.eat_lookahead(TokenType::StringLiteral);
				Ok(ExprNode::init_as_variable(current_lexeme))
			}
			TokenType::MinusOperator => {
				self.eat_lookahead(TokenType::MinusOperator);
				// TODO: Fix this shit, we're appending contextual information in the parser
				// as if it is scanning... dreadful.
				let current_lexeme = format!("-{}", self.get_token().get_lexeme());
				match self.get_lookahead() {
					TokenType::IntegerLiteral => {
						self.eat_lookahead(TokenType::IntegerLiteral);
						Ok(ExprNode::init_as_literal(
							current_lexeme,
							DataType::IntegerType,
						))
					}
					TokenType::FloatLiteral => {
						self.eat_lookahead(TokenType::FloatLiteral);
						Ok(ExprNode::init_as_literal(
							current_lexeme,
							DataType::FloatType,
						))
					}
					TokenType::Identifier => {
						self.eat_lookahead(TokenType::IntegerLiteral);
						Ok(ExprNode::init_as_variable(current_lexeme))
					}
					_ => Err(self.get_static_compile_error(
						"Negative subexpression not matched here",
					)),
				}
			}
			_ => Err(self.get_static_compile_error(
				"Subexpression did not start with id, left paren, or numeric constant",
			)),
		}
	}
}
